pkg/pgmodel/ingestor/ingestor.go

```diff
@@ -52,12 +57,73 @@ func NewPgxIngestorForTests(conn pgxconn.PgxConn) (*DBIngestor, error) {
 //     tts the []Timeseries to insert
 //     req the WriteRequest backing tts. It will be added to our WriteRequest
 //         pool when it is no longer needed.
-func (ingestor *DBIngestor) Ingest(r *prompb.WriteRequest) (uint64, error) {
-	var (
-		err       error
-		totalRows uint64
+func (ingestor *DBIngestor) Ingest(r *prompb.WriteRequest) (numSamples uint64, numMetadata uint64, err error) {
+	defer func() {
+		// WriteRequests can contain pointers into the original buffer we deserialized
+		// them out of, and can be quite large in and of themselves. In order to prevent
+		// memory blowup, and to allow faster deserializing, we recycle the WriteRequest
+		// here, allowing it to be either garbage collected or reused for a new request.
+		// In order for this to work correctly, any data we wish to keep using (e.g.
+		// samples) must no longer be reachable from req.
+		FinishWriteRequest(r)
+	}()
+	switch numTs, numMeta := len(r.Timeseries), len(r.Metadata); {
+	case numTs == 0 && numMeta == 0:
+		return 0, 0, nil
+	case numTs > 0 && numMeta == 0:
+		// Write request contains only time-series.
+		n, err := ingestor.ingestSamples(r)
+		return n, 0, err
+	case numMeta > 0 && numTs == 0:
+		// Write request contains only metadata.
+		n, err := ingestor.ingestMetadata(r)
+		return 0, n, err
+	default:
+		// Write request contains both samples and metadata, hence we ingest concurrently.
+		type result struct {
+			id      int8
+			numRows uint64
+			err     error
+		}
+		res := make(chan result, 2)
```

Comment:
I don't think you need a buffered channel here. Best to avoid them unless there is a clear need for them.